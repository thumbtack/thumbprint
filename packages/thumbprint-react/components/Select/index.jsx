import React from 'react';
import PropTypes from 'prop-types';
import * as tokens from '@thumbtack/thumbprint-tokens';
import classNames from 'classnames';
import styles from './index.module.scss';
import { NavigationCaretDownSmall } from '../../icons/index.jsx';

const iconColor = {
    disabled: tokens.tpColorGray,
    default: tokens.tpColorBlack,
    error: tokens.tpColorRed,
};

const getUIState = ({ isDisabled, hasError }) => {
    if (isDisabled) {
        return 'disabled';
    }

    if (hasError) {
        return 'error';
    }

    return 'default';
};

const Select = ({
    children,
    dataTest,
    hasError,
    id,
    isDisabled,
    isFullWidth,
    isRequired,
    name,
    onChange,
    onFocus,
    onBlur,
    onClick,
    size,
    value,
}) => {
    const uiState = getUIState({ isDisabled, hasError });
    return (
        <div
            className={classNames({
                [styles.root]: true,
                [styles.rootWidthFull]: isFullWidth,
            })}
        >
            <select
                className={classNames({
                    [styles.select]: true,
                    [styles.selectStateDisabled]: uiState === 'disabled',
                    [styles.selectStateError]: uiState === 'error',
                    [styles.selectStateDefault]: uiState === 'default',
                    [styles.selectSizeSmall]: size === 'small',
                    [styles.selectSizeLarge]: size === 'large',
                })}
                id={id}
                disabled={isDisabled}
                required={isRequired}
                value={value}
                onClick={() => onClick && onClick()}
                onChange={e => onChange(e.target.value, e)}
                onFocus={onFocus}
                onBlur={onBlur}
                data-test={dataTest}
                name={name}
            >
                {children}
            </select>

            <NavigationCaretDownSmall className={styles.caret} fill={iconColor[uiState]} />
        </div>
    );
};

Select.propTypes = {
    /**
     * A collection of [HTML `<option>` elements](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/option).
     */
    children: PropTypes.node,
    /**
     * Adds a HTML `id` attribute to the select. This is used for linking the HTML with a
     * [Label](/components/label/react/).
     */
    id: PropTypes.string,
    /**
     * Visually and functionally disables the select dropdown.
     */
    isDisabled: PropTypes.bool,
    /**
     * Adds `required` HTML attribute to element, indicating that an option with a non-empty string
     * value must be selected.
     */
    isRequired: PropTypes.bool,
    /**
     * Makes the radio and text color red.
     */
    hasError: PropTypes.bool,
    /**
     * Changes the selectâ€™s font-size, height, and padding.
     */
    size: PropTypes.oneOf(['small', 'large']),
    /**
     * Set the `<select>`'s width to `100%` as opposed to the default behavior
     * of only taking up the necessary width.
     */
    isFullWidth: PropTypes.bool,
    /**
     * The `value` of the `<option>` that is selected. See [React documentation
     * on `<select>` and controlled components](https://facebook.github.io/react/docs/forms.html#the-select-tag)
     * to learn more.
     */
    value: PropTypes.string.isRequired,
    /**
     * Function that is fired when the value of the select changes.
     */
    onClick: PropTypes.func,
    /**
     * A function that is fired when the value of the select changes. The
     * new `value` is passed to the function.
     */
    onChange: PropTypes.func.isRequired,
    /**
     * Fires when the select receives focus.
     */
    onFocus: PropTypes.func,
    /**
     * Fires when the select loses focus.
     */
    onBlur: PropTypes.func,
    /**
     * A selector hook into the React component for use in automated testing environments.
     */
    dataTest: PropTypes.string,
    /**
     * Adds `name` HTML attribute to element, indicating the property name associated with the selected value.
     */
    name: PropTypes.string,
};

Select.defaultProps = {
    children: undefined,
    id: undefined,
    isDisabled: false,
    isRequired: false,
    hasError: false,
    isFullWidth: false,
    size: 'large',
    onClick: undefined,
    onFocus: undefined,
    onBlur: undefined,
    dataTest: undefined,
    name: undefined,
};

export default Select;
