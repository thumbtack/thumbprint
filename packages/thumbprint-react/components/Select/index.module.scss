@import '~@thumbtack/thumbprint-tokens/dist/scss/_index';

.root {
    display: inline-block;
    position: relative;

    &WidthFull {
        width: 100%;
    }
}

.select {
    // Vendor prefixes are needed to work around a bug in Gatsby, the framework that powers our
    // documentation site, so that the component doesn't appear broken on our documentation.
    // https://github.com/thumbtack/thumbprint-archive/issues/899
    // https://github.com/gatsbyjs/gatsby/issues/1347
    -webkit-appearance: none;
    -moz-appearance: none;
    appearance: none;

    border: 1px solid;
    font-family: $tp-font-family__base;
    border-radius: $tp-border-radius__base;
    outline: none;
    vertical-align: middle;
    background-repeat: no-repeat;
    background-size: auto 18px;
    padding-left: 14px;
    padding-right: 54px;
    width: 100%;

    &:focus {
        border-color: $tp-color__blue;
    }

    &::-ms-expand {
        display: none;
    }

    &:-moz-focusring {
        color: transparent;
        text-shadow: 0 0 0 $tp-color__black;
    }

    &State {
        &Disabled {
            background-color: $tp-color__gray-200;
            border-color: $tp-color__gray;
            color: $tp-color__gray;
        }

        &Error {
            background-color: $tp-color__white;
            border-color: $tp-color__red;
            color: $tp-color__red;

            &:focus {
                border-color: $tp-color__red;
            }
        }

        &Default {
            background-color: $tp-color__white;
            border-color: $tp-color__gray;
            color: $tp-color__black;
        }
    }

    &Size {
        &Small {
            height: 40px;
            font-size: $tp-font__body__2__size;
        }

        &Large {
            height: 52px;
            font-size: $tp-font__title__5__size;
        }
    }
}

.caret {
    position: absolute;
    top: 0;
    right: 16px;
    bottom: 0;
    margin-top: auto;
    margin-bottom: auto;
    pointer-events: none;
}
