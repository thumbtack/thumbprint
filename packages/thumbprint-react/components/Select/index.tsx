import React from 'react';
import * as tokens from '@thumbtack/thumbprint-tokens';
import classNames from 'classnames';
import styles from './index.module.scss';
import { NavigationCaretDownSmall } from '../../icons/index.jsx';

const iconColor = {
    disabled: tokens.tpColorGray,
    default: tokens.tpColorBlack,
    error: tokens.tpColorRed,
};

const getUIState = ({
    isDisabled,
    hasError,
}: {
    isDisabled: boolean;
    hasError: boolean;
}): 'disabled' | 'error' | 'default' => {
    if (isDisabled) {
        return 'disabled';
    }

    if (hasError) {
        return 'error';
    }

    return 'default';
};

interface PropTypes {
    /**
     * A collection of [HTML `<option>` elements](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/option).
     */
    children?: React.ReactNode;
    /**
     * Adds a HTML `id` attribute to the select. This is used for linking the HTML with a
     * [Label](/components/label/react/).
     */
    id?: string;
    /**
     * Visually and functionally disables the select dropdown.
     */
    isDisabled?: boolean;
    /**
     * Adds `required` HTML attribute to element, indicating that an option with a non-empty string
     * value must be selected.
     */
    isRequired?: boolean;
    /**
     * Makes the radio and text color red.
     */
    hasError?: boolean;
    /**
     * Changes the selectâ€™s font-size, height, and padding.
     */
    size?: 'small' | 'large';
    /**
     * Set the `<select>`'s width to `100%` as opposed to the default behavior
     * of only taking up the necessary width.
     */
    isFullWidth?: boolean;
    /**
     * The `value` of the `<option>` that is selected. See [React documentation
     * on `<select>` and controlled components](https://facebook.github.io/react/docs/forms.html#the-select-tag)
     * to learn more.
     */
    value: string;
    /**
     * Function that is fired when the value of the select changes.
     */
    onClick?: () => void;
    /**
     * A function that is fired when the value of the select changes. The
     * new `value` is passed to the function.
     */
    onChange: (value: string, event: React.ChangeEvent<HTMLSelectElement>) => void;
    /**
     * Fires when the select receives focus.
     */
    onFocus?: () => void;
    /**
     * Fires when the select loses focus.
     */
    onBlur?: () => void;
    /**
     * A selector hook into the React component for use in automated testing environments.
     */
    dataTest?: string;
    /**
     * Adds `name` HTML attribute to element, indicating the property name associated with the
     * selected value.
     */
    name?: string;
}

export default function Select({
    children,
    dataTest,
    hasError = false,
    id,
    isDisabled = false,
    isFullWidth = false,
    isRequired = false,
    name,
    onChange,
    onFocus,
    onBlur,
    onClick = (): void => {},
    size = 'large',
    value,
}: PropTypes): JSX.Element {
    const uiState = getUIState({ isDisabled, hasError });

    return (
        <div
            className={classNames({
                [styles.root]: true,
                [styles.rootWidthFull]: isFullWidth,
            })}
        >
            <select
                className={classNames({
                    [styles.select]: true,
                    [styles.selectStateDisabled]: uiState === 'disabled',
                    [styles.selectStateError]: uiState === 'error',
                    [styles.selectStateDefault]: uiState === 'default',
                    [styles.selectSizeSmall]: size === 'small',
                    [styles.selectSizeLarge]: size === 'large',
                })}
                id={id}
                disabled={isDisabled}
                required={isRequired}
                value={value}
                onClick={onClick}
                onChange={(event): void => onChange(event.target.value, event)}
                onFocus={onFocus}
                onBlur={onBlur}
                data-test={dataTest}
                name={name}
            >
                {children}
            </select>

            <NavigationCaretDownSmall className={styles.caret} fill={iconColor[uiState]} />
        </div>
    );
}
