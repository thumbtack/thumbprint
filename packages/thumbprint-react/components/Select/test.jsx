import React from 'react';
import { noop } from 'lodash';
import { mount, render } from 'enzyme';
import Select from './index';

test('adds `disabled` attribute', () => {
    const wrapper = mount(<Select isDisabled onChange={noop} value="goose" />);
    expect(wrapper.find('select').prop('disabled')).toBeTruthy();
    expect(wrapper).toMatchSnapshot();
});

test('adds `id` attribute', () => {
    const wrapper = mount(<Select id="Goose" onChange={noop} value="goose" />);
    expect(wrapper.find('select').prop('id')).toBe('Goose');
    expect(wrapper).toMatchSnapshot();
});

test('adds `required` attribute', () => {
    const wrapper = mount(<Select isRequired onChange={noop} value="goose" />);
    expect(wrapper.find('select').prop('required')).toBeTruthy();
    expect(wrapper).toMatchSnapshot();
});

test('renders an error state with `hasError`', () => {
    const wrapper = mount(<Select hasError onChange={noop} value="goose" />);
    expect(wrapper).toMatchSnapshot();
});

test('renders options that are passed in as `children`', () => {
    const wrapper = mount(
        <Select onChange={noop} value="goose">
            <option>New York</option>
            <option>California</option>
        </Select>,
    );

    expect(wrapper.find('option')).toHaveLength(2);
    expect(wrapper).toMatchSnapshot();
});

test('uses `value` as the selected option', () => {
    const wrapper = render(
        <Select value="ca" onChange={jest.fn}>
            <option value="ny">New York</option>
            <option value="ca">California</option>
        </Select>,
    );
    expect(wrapper.find('select [selected]').val()).toBe('ca');
    expect(wrapper).toMatchSnapshot();
});

test('calls `onChange` function and supplies new value and event', () => {
    const onChange = jest.fn();

    const wrapper = mount(<Select onChange={onChange} value="goose" />);
    const select = wrapper.find('select');

    const event1 = { target: { value: 'He' } };
    select.simulate('change', event1);
    expect(onChange).toHaveBeenCalledTimes(1);
    expect(onChange).toHaveBeenLastCalledWith(event1.target.value, expect.objectContaining(event1));

    const event2 = { target: { value: 'Hello' } };
    select.simulate('change', event2);
    expect(onChange).toHaveBeenCalledTimes(2);
    expect(onChange).toHaveBeenLastCalledWith(event2.target.value, expect.objectContaining(event2));
});

test('calls `onBlur` function when supplied', () => {
    const onBlur = jest.fn();

    const wrapper = mount(<Select onBlur={onBlur} onChange={jest.fn} value="goose" />);

    wrapper.find('select').simulate('blur');
    expect(onBlur).toHaveBeenCalledTimes(1);
});

test('calls `onFocus` function when supplied', () => {
    const onFocus = jest.fn();

    const wrapper = mount(<Select onFocus={onFocus} onChange={jest.fn} value="goose" />);

    wrapper.find('select').simulate('focus');
    expect(onFocus).toHaveBeenCalledTimes(1);
});

test('calls `onClick` function when clicked on', () => {
    const onClick = jest.fn();

    const wrapper = mount(<Select onClick={onClick} onChange={noop} value="goose" />);

    wrapper.find('select').simulate('click');
    expect(onClick).toHaveBeenCalledTimes(1);
});

test('adds `dataTest` prop', () => {
    const wrapperA = mount(<Select dataTest="Duck" onChange={noop} value="goose" />);
    expect(wrapperA.find('select').prop('data-test')).toEqual('Duck');
    expect(wrapperA).toMatchSnapshot();
});

test('adds `name` HTML attribute', () => {
    const wrapperA = mount(<Select name="duck" onChange={noop} value="goose" />);
    expect(wrapperA.find('select').prop('name')).toEqual('duck');
    expect(wrapperA).toMatchSnapshot();
});
