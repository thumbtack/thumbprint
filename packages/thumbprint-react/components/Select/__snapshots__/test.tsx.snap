// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adds \`dataTest\` prop 1`] = `
<Select
  dataTest="Duck"
  onChange={[Function]}
  value="goose"
>
  <div
    className="root"
  >
    <select
      className="select selectStateDefault selectSizeLarge"
      data-test="Duck"
      disabled={false}
      onChange={[Function]}
      onClick={[Function]}
      required={false}
      value="goose"
    />
    <NavigationCaretDownSmall
      className="caret"
      fill="#2f3033"
      height="18"
      viewBox="0 0 18 18"
      width="18"
      xmlns="http://www.w3.org/2000/svg"
    >
      <svg
        className="caret"
        fill="#2f3033"
        height="18"
        viewBox="0 0 18 18"
        width="18"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M14.646 6.764L9 13 3.311 6.725a1 1 0 0 1 1.342-1.482L9 10l4.285-4.699c.2-.187.435-.301.715-.301a1 1 0 0 1 1 1c0 .306-.151.537-.354.764z"
          fillRule="evenodd"
        />
      </svg>
    </NavigationCaretDownSmall>
  </div>
</Select>
`;

exports[`adds \`disabled\` attribute 1`] = `
<Select
  isDisabled={true}
  onChange={[Function]}
  value="goose"
>
  <div
    className="root"
  >
    <select
      className="select selectStateDisabled selectSizeLarge"
      disabled={true}
      onChange={[Function]}
      onClick={[Function]}
      required={false}
      value="goose"
    />
    <NavigationCaretDownSmall
      className="caret"
      fill="#d3d4d5"
      height="18"
      viewBox="0 0 18 18"
      width="18"
      xmlns="http://www.w3.org/2000/svg"
    >
      <svg
        className="caret"
        fill="#d3d4d5"
        height="18"
        viewBox="0 0 18 18"
        width="18"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M14.646 6.764L9 13 3.311 6.725a1 1 0 0 1 1.342-1.482L9 10l4.285-4.699c.2-.187.435-.301.715-.301a1 1 0 0 1 1 1c0 .306-.151.537-.354.764z"
          fillRule="evenodd"
        />
      </svg>
    </NavigationCaretDownSmall>
  </div>
</Select>
`;

exports[`adds \`id\` attribute 1`] = `
<Select
  id="Goose"
  onChange={[Function]}
  value="goose"
>
  <div
    className="root"
  >
    <select
      className="select selectStateDefault selectSizeLarge"
      disabled={false}
      id="Goose"
      onChange={[Function]}
      onClick={[Function]}
      required={false}
      value="goose"
    />
    <NavigationCaretDownSmall
      className="caret"
      fill="#2f3033"
      height="18"
      viewBox="0 0 18 18"
      width="18"
      xmlns="http://www.w3.org/2000/svg"
    >
      <svg
        className="caret"
        fill="#2f3033"
        height="18"
        viewBox="0 0 18 18"
        width="18"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M14.646 6.764L9 13 3.311 6.725a1 1 0 0 1 1.342-1.482L9 10l4.285-4.699c.2-.187.435-.301.715-.301a1 1 0 0 1 1 1c0 .306-.151.537-.354.764z"
          fillRule="evenodd"
        />
      </svg>
    </NavigationCaretDownSmall>
  </div>
</Select>
`;

exports[`adds \`name\` HTML attribute 1`] = `
<Select
  name="duck"
  onChange={[Function]}
  value="goose"
>
  <div
    className="root"
  >
    <select
      className="select selectStateDefault selectSizeLarge"
      disabled={false}
      name="duck"
      onChange={[Function]}
      onClick={[Function]}
      required={false}
      value="goose"
    />
    <NavigationCaretDownSmall
      className="caret"
      fill="#2f3033"
      height="18"
      viewBox="0 0 18 18"
      width="18"
      xmlns="http://www.w3.org/2000/svg"
    >
      <svg
        className="caret"
        fill="#2f3033"
        height="18"
        viewBox="0 0 18 18"
        width="18"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M14.646 6.764L9 13 3.311 6.725a1 1 0 0 1 1.342-1.482L9 10l4.285-4.699c.2-.187.435-.301.715-.301a1 1 0 0 1 1 1c0 .306-.151.537-.354.764z"
          fillRule="evenodd"
        />
      </svg>
    </NavigationCaretDownSmall>
  </div>
</Select>
`;

exports[`adds \`required\` attribute 1`] = `
<Select
  isRequired={true}
  onChange={[Function]}
  value="goose"
>
  <div
    className="root"
  >
    <select
      className="select selectStateDefault selectSizeLarge"
      disabled={false}
      onChange={[Function]}
      onClick={[Function]}
      required={true}
      value="goose"
    />
    <NavigationCaretDownSmall
      className="caret"
      fill="#2f3033"
      height="18"
      viewBox="0 0 18 18"
      width="18"
      xmlns="http://www.w3.org/2000/svg"
    >
      <svg
        className="caret"
        fill="#2f3033"
        height="18"
        viewBox="0 0 18 18"
        width="18"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M14.646 6.764L9 13 3.311 6.725a1 1 0 0 1 1.342-1.482L9 10l4.285-4.699c.2-.187.435-.301.715-.301a1 1 0 0 1 1 1c0 .306-.151.537-.354.764z"
          fillRule="evenodd"
        />
      </svg>
    </NavigationCaretDownSmall>
  </div>
</Select>
`;

exports[`renders an error state with \`hasError\` 1`] = `
<Select
  hasError={true}
  onChange={[Function]}
  value="goose"
>
  <div
    className="root"
  >
    <select
      className="select selectStateError selectSizeLarge"
      disabled={false}
      onChange={[Function]}
      onClick={[Function]}
      required={false}
      value="goose"
    />
    <NavigationCaretDownSmall
      className="caret"
      fill="#ff5a5f"
      height="18"
      viewBox="0 0 18 18"
      width="18"
      xmlns="http://www.w3.org/2000/svg"
    >
      <svg
        className="caret"
        fill="#ff5a5f"
        height="18"
        viewBox="0 0 18 18"
        width="18"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M14.646 6.764L9 13 3.311 6.725a1 1 0 0 1 1.342-1.482L9 10l4.285-4.699c.2-.187.435-.301.715-.301a1 1 0 0 1 1 1c0 .306-.151.537-.354.764z"
          fillRule="evenodd"
        />
      </svg>
    </NavigationCaretDownSmall>
  </div>
</Select>
`;

exports[`renders options that are passed in as \`children\` 1`] = `
<Select
  onChange={[Function]}
  value="goose"
>
  <div
    className="root"
  >
    <select
      className="select selectStateDefault selectSizeLarge"
      disabled={false}
      onChange={[Function]}
      onClick={[Function]}
      required={false}
      value="goose"
    >
      <option>
        New York
      </option>
      <option>
        California
      </option>
    </select>
    <NavigationCaretDownSmall
      className="caret"
      fill="#2f3033"
      height="18"
      viewBox="0 0 18 18"
      width="18"
      xmlns="http://www.w3.org/2000/svg"
    >
      <svg
        className="caret"
        fill="#2f3033"
        height="18"
        viewBox="0 0 18 18"
        width="18"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M14.646 6.764L9 13 3.311 6.725a1 1 0 0 1 1.342-1.482L9 10l4.285-4.699c.2-.187.435-.301.715-.301a1 1 0 0 1 1 1c0 .306-.151.537-.354.764z"
          fillRule="evenodd"
        />
      </svg>
    </NavigationCaretDownSmall>
  </div>
</Select>
`;

exports[`uses \`value\` as the selected option 1`] = `
<div
  class="root"
>
  <select
    class="select selectStateDefault selectSizeLarge"
  >
    <option
      value="ny"
    >
      New York
    </option>
    <option
      selected=""
      value="ca"
    >
      California
    </option>
  </select>
  <svg
    class="caret"
    fill="#2f3033"
    height="18"
    viewBox="0 0 18 18"
    width="18"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M14.646 6.764L9 13 3.311 6.725a1 1 0 0 1 1.342-1.482L9 10l4.285-4.699c.2-.187.435-.301.715-.301a1 1 0 0 1 1 1c0 .306-.151.537-.354.764z"
      fill-rule="evenodd"
    />
  </svg>
</div>
`;
